bin/Macronify/Main.cpp

```diff
@@ -0,0 +1,734 @@
+                    return is_legal;
+                } else if (is_rcu_dereference) {
+                    exp.getExpansion().walk([&](Op *cur) {
+                        if (auto param = mlir::dyn_cast<MP>(cur)) {
+                            auto param_name = param.getParameterName();
+                            if (param_name == "p") {
+                                rcu_dereference_to_p.insert({ op, cur });
+                            }
+                        }});
+                    bool found_p = rcu_dereference_to_p.contains(op),
+                        is_legal = !found_p;
+                    return is_legal;
+                } else if (is_smp_mb) {
+                    return false;
+                } else {
+                    return true;
+                }
+            });
+        trg.addDynamicallyLegalOp<vast::hl::ForOp>(
+            [](Op *op) {
+                auto for_op = mlir::dyn_cast<vast::hl::ForOp>(op);
+                using MP = macroni::macroni::MacroParameter;
+                for_op.getCondRegion().walk(
+                    [&](Op *op2) {
+                        if (auto param_op = mlir::dyn_cast<MP>(op2)) {
+                            if (param_op.getParameterName() == "pos") {
+                                list_for_each_to_pos[op] = op2;
+                            } else if (param_op.getParameterName() == "head") {
+                                list_for_each_to_head[op] = op2;
+                            }
+                        }
+                    }
+                );
+                return !(list_for_each_to_pos.contains(op) &&
+                         list_for_each_to_head.contains(op));
+            });
+        trg.addDynamicallyLegalOp<vast::hl::CallOp>(
+            [](Op *op) {
+                using CO = vast::hl::CallOp;
+                CO call_op = mlir::dyn_cast<CO>(op);
+                llvm::StringRef name = call_op.getCalleeAttr().getValue();
+                if (rcu_read_unlock != name) {
+                    return true;
+                }
+                llvm::APInt unlock_level = get_lock_level(op);
+                for (auto cur = op; cur; cur = cur->getPrevNode()) {
+                    if (CO call_op = mlir::dyn_cast<CO>(cur)) {
+                        name = call_op.getCalleeAttr().getValue();
+                        if (rcu_read_lock == name) {
+                            llvm::APInt lock_level = get_lock_level(cur);
+                            if (unlock_level == lock_level) {
+                                unlock_to_lock[op] = cur;
+                                break;
+                            }
+                        }
+                    }
+                }
+                return !unlock_to_lock.contains(op);
+            });
+        trg.markUnknownOpDynamicallyLegal([](Op *op) { return true;});
+
+        mlir::RewritePatternSet patterns(&*mctx);
+        patterns.add<macroni::macro_expansion_to_get_user>(patterns.getContext());
```

Comment:
As a C++ exercise you can try to write a wrapper that accepts variadic number of patterns.

Also, I am pretty sure `add< ... >` already supports that.